### DISCUSSION TITLE
Dynamic Module Federation: singleton library instantiated 2x

### DISCUSSION THREAD

**Discussion by jarecsni** [webpack/webpack#18769] [Q&A] [answered]
In a project with a host and a remote app, I noticed that the npm library I'm developing (it's a framework which is used both by host and remote apps), is bundled by the remote as well. I tried all the various settings, etc, to no avail. If I set a debug breakpoint I can see that Singleton classes in this library instantiated 2x first by the host app, and then by the remote. 

My issue is the same as described here: https://stackoverflow.com/questions/65467233/wp5-module-federation-singleton-instantiated-multiple-times?rq=2

Hoping someone can point me towards some documentation or better yet, working example. (I did look at for example the repo with tons of federation examples https://github.com/module-federation/module-federation-examples however couldn't find any solution to the problem there.

**Comment by alexander-akait** [comment]
You can open an issue in - https://github.com/module-federation/module-federation-examples for examples, shortly - you need to use `shared` and `singleton` - https://webpack.js.org/plugins/module-federation-plugin/#singleton and https://webpack.js.org/plugins/module-federation-plugin/#additional-examples

**Reply by jarecsni** [reply]
Thanks for your reply! I have an open discussion topic on this issue in the module-federation-examples repo: https://github.com/module-federation/module-federation-examples/discussions/4270 unfortunately it didn't attract any attention so far. I do use shared and singleton, but they seem to be ineffective. I'm really at a loss as to what's going on it's kind of the point of module federation right - so either I'm missing something really fundamental, or this is not meant to be working at all :D

I've also opened a question on Stackoverflow (with a bounty of 200): https://stackoverflow.com/questions/78985151/webpack-5-module-federation-shared-library-issue - again, 0 attention so far.

**Reply by alexander-akait** [reply]
Feel free to open an issue, also you can create a small reproducible test repo using github and I will look

**Reply by jarecsni** [reply]
Thanks for your kind offer, I'll put together a test repo over the weekend, will post it here (and also as you suggested open an issue as well).

**Reply by jarecsni** [reply, marked as answer]
Finally I got a break through. I went through my problem with a colleague who spotted that webpack doesn't emit any message for my library ('consume shared module / provide shared module') whereas such a message was there for react and other libraries. The library project didn't properly support ESM which caused webpack to exclude it (without an explicit error message which isn't ideal). Fixing the library to be CJS+ESM fixed the problem, only a single instance is loaded, Singletons finally work - happy days :)
